### PR TITLE
feat: add Button component

### LINKED ISSUE #107: Feature: Add `Button` to component
## Issue Type

- [ ] Bug Report
- [x] Feature Request
- [x] Enhancement
- [ ] Other (Please specify)

## Description

We are planning to introduce a new `Button` component to our project, and to ensure its versatility, we'd like to implement it using `forwardRef`. This component should be added to the file `components/primitives/Button/button.tsx`.

**Specifications:**

The `Button` component should meet the following specifications:

- **Variants**: It should support multiple button variants, including primary, secondary, and danger buttons.
- **Sizes**: Different button sizes should be available, such as small, medium, and large.
- **Disabled State**: The component should handle a disabled state for buttons.
- **Events**: It should support common button events like `onClick`.

**Design Guidelines:**

The `Button` component's design should align with our project's UI/UX guidelines, maintaining consistency with the established color schemes. Please refer to the Figma design for visual guidelines: [Figma Design](https://www.figma.com/file/WgmdhEtnF0C90qPuvvm3OS/AbegHelp-(Copy)?type=design&node-id=316-652&mode=design&t=cnaEw6puMBiT7sqE-4).

**Implementation Details:**

To fulfill this request, you should:

1. Implement the `Button` component using `forwardRef`.
2. Create a TypeScript file named `button.tsx` in the `components/primitives/Button` directory.
3. Develop the `Button` component with support for different variants, sizes, and states as specified.
4. Implement the necessary event handling (e.g., `onClick`) for the button.
5. Ensure the component includes a spread props functionality to allow other props.

**Steps to Take:**

1. Implement the `Button` component using `forwardRef`.
2. Create the `button.tsx` file in the `components/primitives/Button` directory.
3. Develop the `Button` component with the specified design guidelines and features.
4. Ensure the component's behavior adheres to the requirements for variants, sizes, and events.
5. Add a spread props functionality to the `Button` component to enable the use of other props.

Here's a basic code snippet to get you started:

```jsx
import React, { forwardRef } from 'react';

interface ButtonProps {
  variant: 'primary' | 'secondary' | 'danger';
  size: 'sm' | 'base' | 'lg';
  disabled?: boolean;
  onClick?: () => void;
  label: string;
}

const Button = forwardRef<HTMLButtonElement, ButtonProps>(({ variant, size, disabled, onClick, label, ...props }, ref) => {
  return (
    <button ref={ref} className={classNames} onClick={onClick} disabled={disabled} {...props}>
      {label}
    </button>
  );
});

Button.displayName = 'Button';

export default Button;
```

**Expected Outcome:**

The expected outcome is the successful implementation of the `Button` component using `forwardRef`, along with the addition of the spread props functionality. This implementation will provide a versatile and reusable button component for various user interactions within our project.

**Notte:**

If you have any questions or need further details, please feel free to ask.

### PR DESCRIPTION
**Pull request by jemmycodes**
## Description

Created a reusable Button component 

## Related Issue

- Fixes #107 

## Contribution Guidelines

Before submitting this pull request, please review our [Contribution Guidelines](https://github.com/abeg-help/frontend/blob/dev/CONTRIBUTING.md) to understand how to contribute to this project.

## Checklist

- [x] I have reviewed the Contribution Guidelines linked above.
- [x] I have tested my changes thoroughly and ensured that all existing tests pass.
- [x] I have provided clear and concise commit messages.
- [x] I have updated the project's documentation as necessary.

## Additional context (if needed)

https://github.com/abeg-help/frontend/assets/110843645/0b18907b-fed7-41b3-9917-6f76d4444e2c

### CODE REVIEW

File: components/primitives/Button/Button.tsx [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import { forwardRef } from "react";
+
+interface ButtonProps {
+  className: string;
+  disabled?: boolean;
+  onClick?: () => void;
+  children: React.ReactNode;
+  size: "sm" | "base" | "lg";
+  variant: "primary" | "secondary" | "danger";
```
Comment (babblebey) [inline components/primitives/Button/Button.tsx]: I don't see any logic that uses these variation in props.

Take this case for example.

If I use this button like this 

```js
<Button variant="primary" size="sm" />
```

or this way

```js
<Button variant="danger" size="lg" />
```

I should get different looking/styled buttons, but with this current setup, I wouldn't.

Have you tried the button with different props??

### Rightfully!

when I set the `variant` prop to say `primary` there should be a certain className that should be applied to the button.

Also when I select a `lg` as `size` for the button, The button size should be larger than when it's at `sm` or `base` 😉

**Comment by 289Volts** [comment]
Well done big dev. However, the component is not yet accessible as it does not have the functionality to receive accessible attributes like aria. 

Please add an extension of HTML element attributes to the component. 

Also, I think size and variants props should be optional as the button component might sometimes not need one of the two props

**Review by babblebey — CHANGES_REQUESTED** [review]
Great start @jemmycodes 

You're just missing logic implementation for certain props

**Comment by jemmycodes** [comment]
> Well done big dev. However, the component is not yet accessible as it does not have the functionality to receive accessible attributes like aria.
> 
> Please add an extension of HTML element attributes to the component.
> 
> Also, I think size and variants props should be optional as the button component might sometimes not need one of the two props

Right, thanks for the feedback. This should be role, aria-label and aria-disabled, right? Any other i need to add?

File: components/primitives/Button/Button.tsx (hunk 1 above)
Comment (jemmycodes) [reply components/primitives/Button/Button.tsx]: Thank you, I was thinking I'll need a style guide or design to implement this?

**Comment by 289Volts** [comment]
> Right, thanks for the feedback. This should be role, aria-label and aria-disabled, right? Any other i need to add?


It's gonna be more than this but you got it. I think there would be a way to spread the HTML element attributes on the element so that if you pass it as prop, it should work. Better still, check w3schools button element and check the global attribute it receives. Let's see if this is feasible

**Comment by jemmycodes** [comment]
> It's gonna be more than this but you got it. I think there would be a way to spread the HTML element attributes on the element so that if you pass it as prop, it should work. Better still, check w3schools button element and check the global attribute it receives. Let's see if this is feasible

Sorry, I'm a little bit confused here. The button component already accepts `...props`, so the HTML accessibility attributes can be added when calling the button element by props. Do you prefer me adding them directly to the button?

File: components/primitives/Button/Button.tsx (hunk 1 above)
Comment (babblebey) [reply components/primitives/Button/Button.tsx]: Just go ahead and make something up.

In regards the color, do check out the configured `abeg-green`, `abeg-error` and others in the `tailwind.config.ts` file; Then the sizing, kindly compose anything you can find for now, these are some tiny changes we can always address when the style guide become available in the future

😉

**Comment by ryan-zayne** [comment]
> Sorry, I'm a little bit confused here. The button component already accepts `...props`, so the HTML accessibility attributes can be added when calling the button element by props. Do you prefer me adding them directly to the button?

There's a way to get around all these without much stress. I'll explain with snippets when you're free 🌝

For now tho, great work🍀

**Review by ryan-zayne — APPROVED** [review]
Superb work!🥂

File: components/primitives/Button/Button.tsx [hunk 2]
```diff
@@ -0,0 +1,50 @@
+import { cn } from "@/lib/utils/cn";
```
Comment (babblebey) [inline components/primitives/Button/Button.tsx]: BLOCKING

Kindly rename the component file to a lowercase format.

Going from `Button.tsx` to `button.tsx`

File: components/primitives/Button/Button.tsx [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    "px-10": size === "lg",
+    "w-full": fullWidth,
+    "bg-abeg-neutral-50 text-white": disabled,
+  };
+
+  return (
+    <button
+      className={cn(
+        "px-5 py-2 rounded-lg text-white text-sm font-medium",
+        buttonClass,
```
Comment (babblebey) [inline components/primitives/Button/Button.tsx]: Impressive what I see here 🤓

But, I'm curious to know how the `className` is resolved from the usage of the `buttonClass` object.

If you got a resource you wanna share that explains this approach, do share or If you want to lesson me too, I'm down. 

🤲🏾

**Review by babblebey — COMMENTED** [review]
Hey @jemmycodes, great stuff once again.

Kindly address the "Blocking" review, and yo do add demo video of the component in its various props to ths PR Description.

It will aid a visual review, you know see the component working with various different props.

File: components/primitives/Button/Button.tsx (hunk 2 above)
Comment (ryan-zayne) [reply components/primitives/Button/Button.tsx]: A little curious to as why the lowercase convention for Components🤔

Is this to help migitate differences between how different operating systems resolve filenames, or is it purely an opinionated approach? Thoughts🤔

File: components/primitives/Button/Button.tsx (hunk 3 above)
Comment (ryan-zayne) [reply components/primitives/Button/Button.tsx]: It uses the varadic object format (which apparently you love so much🌝), which clsx accepts and resolves to the className strings gotten from the object keys, as long as the condition placed as the object values is true at runtime 🤝

Comment (jemmycodes) [reply components/primitives/Button/Button.tsx]: > Impressive what I see here 🤓
> 
> But, I'm curious to know how the `className` is resolved from the usage of the `buttonClass` object.
> 
> If you got a resource you wanna share that explains this approach, do share or If you want to lesson me too, I'm down. 
> 
> 🤲🏾

Yeah I just used the clsx docs to understand the package.

**Review by babblebey — APPROVED** [review]

**Pull request merged**